crates/re_types/src/lib.rs

```diff
@@ -197,27 +194,49 @@ pub trait Loggable {
 
     /// Given an Arrow array, deserializes it into a collection of optional [`Loggable`]s.
     ///
-    /// This will _never_ fail for if the Arrow array's datatype matches the one returned by
+    /// This will _never_ fail if the Arrow array's datatype matches the one returned by
     /// [`Loggable::to_arrow_datatype`].
     /// For the fallible version, see [`Loggable::try_from_arrow_opt`].
     #[inline]
-    fn from_arrow_opt(data: &dyn ::arrow2::array::Array) -> Vec<Option<Self>>
-    where
-        Self: Sized,
-    {
+    fn from_arrow_opt(data: &dyn ::arrow2::array::Array) -> Vec<Option<Self>> {
         Self::try_from_arrow_opt(data).detailed_unwrap()
     }
 
     /// Given an Arrow array, deserializes it into a collection of optional [`Loggable`]s.
     ///
-    /// This will _never_ fail for if the Arrow array's datatype matches the one returned by
+    /// This will _never_ fail if the Arrow array's datatype matches the one returned by
     /// [`Loggable::to_arrow_datatype`].
     /// For the non-fallible version, see [`Loggable::from_arrow_opt`].
+    #[inline]
     fn try_from_arrow_opt(
         data: &dyn ::arrow2::array::Array,
-    ) -> DeserializationResult<Vec<Option<Self>>>
-    where
-        Self: Sized;
+    ) -> DeserializationResult<Vec<Option<Self>>> {
+        Ok(Self::try_from_arrow_iter_item(data)?
+            .map(Self::convert_item_to_self)
+            .collect())
+    }
+
+    /// Given an Arrow array, deserializes it into a iterator of [`Loggable::Item`]s.
+    ///
+    /// Note: mostly for reasons related to typing of trait implementations, the implementor
+    /// of [`Loggable`] may choose an arbitrary iterable [`Loggable::Item`] that  differs from
+    /// the [`Loggable`] itself.
+    ///
+    /// These items can be be converted to an optional [`Loggable`] using [`Loggable::convert_item_to_self`].
+    ///
+    /// This is the base deserialization mechanism that all [`Loggable`] implementors must provide. All other
+    /// conversions above can be generated from this primitive.
+    ///
+    /// This will _never_ fail for if the Arrow array's datatype matches the one returned by
+    /// [`Loggable::to_arrow_datatype`].
+    fn try_from_arrow_iter_item(
```

Comment:
I find this name confusing, it sounds like we're building something from an iterator of arrow data.

How about `try_iter_from_arrow`?